### PR TITLE
Better reference provider

### PR DESCRIPTION
**Pull request by zeroZshadow**
#### Description
Refactor reference providers to no longer require generated classes.
Use `Handle` classes instead of raw `uint` to remember type information.

#### Documentation
* [x] Changelog

### CODE REVIEW

File: test-project/Assets/EditmodeTests/Collections/ReferenceProviderTests.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                world.Connection.CreateEntity(entityId, new EntityTemplate());
+
+                world.Connection.AddComponent(entityId, Position.ComponentId,
+                    new Position.Update {Coords = Coordinates.Zero});
+
+                world.Connection.AddComponent(entityId, Metadata.ComponentId,
+                    new Metadata.Update {EntityType = "EntityWithReferenceComponent"});
+            });
+
+            Assert.AreEqual(1, ReferenceProvider<string>.Count);
```
Comment (jamiebrynes7) [inline test-project/Assets/EditmodeTests/Collections/ReferenceProviderTests.cs:29]: If this assertion fails, the world won't be torn down

File: test-project/Assets/EditmodeTests/Collections/ReferenceProviderTests.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                {
+                    Assert.AreEqual(1, ReferenceProvider<string>.Count);
+                    world.Connection.RemoveComponent(entityId, Metadata.ComponentId);
+                });
+
+                Assert.AreEqual(0, ReferenceProvider<string>.Count);
+            }
+        }
+    }
+}
```
Comment (jamiebrynes7) [inline test-project/Assets/EditmodeTests/Collections/ReferenceProviderTests.cs:59]: Nit: newline

File: workers/unity/Packages/io.improbable.gdk.core/Collections/ReferenceProvider.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            Storage.Add(handle, default);
+
+            return handle;
+        }
+
+        private static T Get(ReferenceHandle handle)
+        {
+            if (!Storage.TryGetValue(handle, out var value))
+            {
+                throw new ArgumentException($"EntityTypeProvider does not contain handle {handle}");
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.gdk.core/Collections/ReferenceProvider.cs:65]:
```suggestion
                throw new ArgumentException($"{typeof(ReferenceProvider<T>)} does not contain handle {handle}");
```

Comment (jamiebrynes7) [reply workers/unity/Packages/io.improbable.gdk.core/Collections/ReferenceProvider.cs:65]: Also should the `{handle}` arg print out the underlying ID?

File: workers/unity/Packages/io.improbable.gdk.core/Collections/ReferenceProvider.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            }
+
+            return value;
+        }
+
+        private static void Set(ReferenceHandle handle, T value)
+        {
+            if (!Storage.ContainsKey(handle))
+            {
+                throw new ArgumentException($"EntityTypeProvider does not contain handle {handle}");
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.gdk.core/Collections/ReferenceProvider.cs:75]:
```suggestion
                throw new ArgumentException($"{typeof(ReferenceProvider<T>)} does not contain handle {handle}");
```

File: workers/unity/Packages/io.improbable.gdk.core/Collections/ReferenceProvider.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        }
+
+        private static ReferenceHandle GetNextHandle()
+        {
+            nextHandleId++;
+
+            while (Storage.ContainsKey(new ReferenceHandle(nextHandleId)))
+            {
+                nextHandleId++;
+            }
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.gdk.core/Collections/ReferenceProvider.cs:93]: Why do we have this busy loop again?

Comment (zeroZshadow) [reply workers/unity/Packages/io.improbable.gdk.core/Collections/ReferenceProvider.cs:93]: We'll never actually run into this, but it's for when you get an overflow.

Comment (jamiebrynes7) [reply workers/unity/Packages/io.improbable.gdk.core/Collections/ReferenceProvider.cs:93]: Ah right

**Pull request merged**